arbnode/espresso_delayed_message_fetcher.go

```diff
@@ -1,291 +1,214 @@
+/*
+backFill fetches all the delayed messages till a `matureL1Block` which is within the saferty tolerance of the rollup
+and stores them in the database
+*/
+func (d *DelayedMessageFetcher) backFill(ctx context.Context) error {
+	// Get the l1 block number based on the read mode
+	matureL1Block, err := d.getL1BlockNumber(ctx)
 	if err != nil {
-		return nil, err
+		log.Error("Error getting l1 block number", "err", err)
+		return err
 	}
-	// if the current L1 block is less than the from block this means no new L1 blocks have been added
-	// since we did the last read, so we can just return nil
-	if currL1 < f.fromBlock {
-		return nil, fmt.Errorf("l1 block number %d is less than from block %d", currL1, f.fromBlock)
-	}
-
-	log.Debug("Current L1 block and from block:", "currL1", currL1, "fromBlock", f.fromBlock)
 
-	startBlock := f.fromBlock
-	endBlock := currL1
-	hasFound := false
-
-	batch := f.db.NewBatch()
-	// Lookup `MessageDelivered` events from the `startBlock` to `startBlock + blocksToRead`
-	// see if any of them have a `message` field that matches the `seqNum` we are looking for
-	for startBlock <= endBlock && !hasFound {
-		from := big.NewInt(0).SetUint64(startBlock)
-		to := big.NewInt(0).SetUint64(startBlock + f.blocksToRead)
-
-		log.Debug("Looking for delayed messages from range", "from", from, "to", to)
-		msgs, err := f.delayedBridge.LookupMessagesInRange(context.Background(), from, to, nil)
-		if err != nil {
-			log.Error("Failed to lookup delayed messages", "err", err)
-			return nil, err
-		}
-		for _, msg := range msgs {
-			seqNum, err := msg.Message.Header.SeqNum()
+	// Get the from block number from the delayed message fetcher
+	// config. Note: Its important in the first read we read from the config
+	// and not the database because the user might want to start reading from a `fromBlock`
+	// which is before the delayed message number stored in the database
+	fromBlock := d.fromBlock
+	log.Info("backfilling delayed messages", "fromBlock", fromBlock, "matureL1Block", matureL1Block)
+	batch := d.db.NewBatch()
+
+	// Loop through the blocks until we reach the matureL1Block
+	for fromBlock < matureL1Block {
+		if (matureL1Block - fromBlock) > d.maxBlocksToRead {
+			// If the difference is greater than the maxBlocksToRead,
+			// then set the endBlock to fromBlock + maxBlocksToRead
+			err := d.getDelayedMessagesInRange(ctx, batch, fromBlock, fromBlock+d.maxBlocksToRead)
 			if err != nil {
-				return nil, err
-			}
-			if seqNum == index {
-				hasFound = true
+				log.Error("failed to get delayed messages in range", "err", err, "fromBlock", fromBlock, "endBlock", fromBlock+d.maxBlocksToRead)
+				return err
 			}
-			err = f.storeDelayedMessage(batch, seqNum, *msg)
+			fromBlock += d.maxBlocksToRead
+		} else {
+			// If the difference is less than the maxBlocksToRead,
+			// then set the endBlock to matureL1Block
+			err := d.getDelayedMessagesInRange(ctx, batch, fromBlock, matureL1Block)
```

Comment:
nit: `getDelayedMessagesInRange` and the following conditional check / `fromBlock` assignment all mirror the positive condition above.  The only thing that changes is the end block being passed.  This conditional can be cleaned up to make it clear which part of the logic changes with the conditional.